src/rpc/rpc_payment.cpp

```diff
@@ -0,0 +1,355 @@
+    info.last_request_timestamp = ts;
+    if (info.credits < payment)
+    {
+      MDEBUG("Not enough credits: " << info.credits << " < " << payment);
+      return false;
+    }
+    info.credits -= payment;
+    add64clamp(info.credits_used, payment);
+    add64clamp(m_credits_used, payment);
+    MDEBUG("client " << client << " paying " << payment << " for " << rpc << ", " << info.credits << " left");
+    credits = info.credits;
+    return true;
+  }
+
+  bool rpc_payment::get_info(const crypto::public_key &client, const std::function<bool(const cryptonote::blobdata&, cryptonote::block&)> &get_block_template, cryptonote::blobdata &hashing_blob, const crypto::hash &top, uint64_t &diff, uint64_t &credits_per_hash_found, uint64_t &credits)
+  {
+    client_info &info = m_client_info[client]; // creates if not found
+    const uint64_t now = time(NULL);
+    bool need_template = top != info.top || now >= info.block_template_update_time + STALE_THRESHOLD;
+    if (need_template)
+    {
+      cryptonote::blobdata extra_nonce("\x42\x42\x42\x42", 4);
+      if (!get_block_template(extra_nonce, info.block))
+        return false;
+      if(!remove_field_from_tx_extra(info.block.miner_tx.extra, typeid(cryptonote::tx_extra_nonce)))
+        return false;
+      extra_nonce = cryptonote::blobdata((const char*)&client, 4);
+      if(!add_extra_nonce_to_tx_extra(info.block.miner_tx.extra, extra_nonce))
+        return false;
+      hashing_blob = get_block_hashing_blob(info.block);
+      info.previous_top = info.top;
+      info.previous_payments = info.payments;
+      info.payments.clear();
+      info.hashing_blob = hashing_blob;
+      info.block_template_update_time = now;
+    }
+    info.top = top;
+    info.update_time = now;
+    hashing_blob = info.hashing_blob;
+    diff = m_diff;
+    credits_per_hash_found = m_credits_per_hash_found;
+    credits = info.credits;
+    return true;
+  }
+
+  bool rpc_payment::submit_nonce(const crypto::public_key &client, uint32_t nonce, const crypto::hash &top, int64_t &error_code, std::string &error_message, uint64_t &credits, crypto::hash &hash, cryptonote::block &block)
+  {
+    client_info &info = m_client_info[client]; // creates if not found
+    MINFO("client " << client << " sends nonce: " << nonce);
+    const bool is_current = top == info.top;
+    std::unordered_set<uint64_t> &payments = is_current ? info.payments : info.previous_payments;
+    if (payments.find(nonce) != payments.end())
+    {
+      MWARNING("Duplicate nonce " << nonce << " from " << (is_current ? "current" : "previous"));
+      ++m_nonces_dupe;
+      ++info.nonces_dupe;
+      info.credits = std::max(info.credits, PENALTY_FOR_DUPLICATE * m_credits_per_hash_found) - PENALTY_FOR_DUPLICATE * m_credits_per_hash_found;
+      error_code = CORE_RPC_ERROR_CODE_DUPLICATE_PAYMENT;
+      error_message = "Duplicate payment";
+      return false;
+    }
+    payments.insert(nonce);
+
+    if (info.hashing_blob.size() < 43)
+    {
```

Comment:
Where does this magic number 43 come from?

Reply:
It's to match:
#define VARIANT1_CHECK() \
  do if (length < 43) \
  { \
    fprintf(stderr, "Cryptonight variant 1 needs at least 43 bytes of data"); \
    _exit(1); \
  } while(0)

The nonce is at bytes 39-43. Though technically I think some of the data before it is variable size (the timestamp IIRC), though will not change size for quite a while.